### DISCUSSION TITLE
Stub leakage?

### DISCUSSION THREAD

**Discussion by bkoconnell** [bkoconnell/datascience-capstone#16] [Ideas / Brainstorming]
@kristinpredeck @Cadamson-prog 

I was looking at the particle_labeled.parquet data the other day, confirming that there are no duplicates. The good news is that all pairings of stub_id + particle_id are unique, so all 2,801,667 observations are unique.

But after class on Tuesday I was thinking about leakage related to Stub IDs.  There are 3,919 different stubs.  If we train the model without considering potential variation across each stub, are we introducing leakage?  Part of me thinks when we do the train/test split, we need to make sure that there are no Stub ID's in the "test" set that also exist in the "train" set to avoid having the model learn stub-specific patterns and then overperform on the test data.

Thoughts?

I know I might be getting a little ahead here. We're still only at preprocessing & feature engineering, but just wanted to get thinking about how to correctly split the data and minimize leakage as we get closer to model exploration.

**Comment by Cadamson-prog** [comment]
I think you're right. 

What if we split out the data by stub_id, not by individual row? That way entire stubs are within the test or training set, not both—which would create the leakage you alluded to.

**Comment by kristinpredeck** [comment]
Yea solid intuition on that. We can just use group-aware splitting (scikit-learn: GroupShuffleSplit or GroupKFold). This isn't traditional leakage in the sense of using future labels or features that are a linear combination of the target or something, but I guess it’s dependency leakage/group leakage. So more like contamination where models can learn shortcuts, especially NNs and test performance can become inflated.
